### PR TITLE
Windows wheels

### LINKED ISSUE #20: Installing jq onto  Windows
The library looks great.  I'm doing development in Python on windows.  And not sure how to install this on Windows as a python library.   I took at the setting up on the Linux side.  And didn't figure out how to deploy this to Windows.  Any links to help? Thanks

### PR DESCRIPTION
**Pull request by du33169**
Added support for building windows wheels (cp37~cp312, win_amd64/win32). Should fix #20, partially?

The reason it cannot be done directly is that the building toolchain used by upstream [jq](https://github.com/jqlang/jq) can only run on a unix-like system. 

Thankfully, [MSYS2](https://www.msys2.org/) provides a compatible environment to do that on windows. However, we cannot build jqpy entirely under MSYS2 along with lib jq due to [some naming and linking issues](https://github.com/mwilliamson/jq.py/issues/20#issuecomment-2118669913) mentioned previously.

Therefore, a possible workflow is to build lib jq under MSYS2 first, and then build the jqpy wheel linking to that (using standard MSVC python distribution, not the python in MSYS2). 

MinGW GCC is used in above process. It has to be noted, that there does exist some possibilities of potential incompatibilities when the wheels are installed to python built with MSVC, but filling the blank of functional wheels for windows users should come with higher priority.

Changes:

1. A makefile `build_jq.mk` for building the jq library under MSYS2. (It should work on other systems as well)
2. Modified `setup.py`, adding support to use a pre-built jq library, and some additional options required on windows.
3. Modified `jq_tests.py` and `jq_old_tests.py`. Extended `expected_strings` for windows in `test_value_error_is_raised_if_program_is_invalid()`. Now the wheel is supposed to pass all tests on windows.
4. Modified github action workflow `build.yml`, adding a job `build_wheel_win` for building windows wheels.
5. Remove python version 3.5 from the workflow job `tests`, as we no longer build wheels for it (deprecated by [cibuildwheel at v2.0.0](https://cibuildwheel.pypa.io/en/stable/changelog/#v200)). And it causes `setup-python` action to fail recently (see [this issue](https://github.com/actions/setup-python/issues/866)).

For review and testing purposes, a full workflow run building wheels for all python versions and arches can be viewed [here](https://github.com/du33169/jq.py/actions/runs/9207257301), with [its built wheels](https://github.com/du33169/jq.py/actions/runs/9207257301/artifacts/1530830013). The latest commit maintains consistency with the original build strategy(only build all wheels on tag push).

Additional Notes:

- The built jq lib files will be cached(per jq version and arch) for speeding up the workflow. Currently this is only enabled(and required) for building windows wheels, but it can also be employed on other systems.
- I am not familiar with tox, so the workflow job named `tests` was not extended for windows. But I do notice that cibuildwheel will also run the tests on built wheels.
- Failed to build wheels for CPython3.6 on windows. There is a workaround that to patch `[python root]\lib\distutils\cygwinccompiler.py`([from here](https://bugs.python.org/issue25251#msg251758)). Currently this is not employed in this PR, and I personally don't recommend it.
- Failed to build wheels for PyPy on windows, reporting `cannot find -lvcruntime140: no such file or directory`. I'm also not familiar with PyPy so this remains unresolved for now.

Acknowledgment:

This PR is initially inspired by @jknockel, who provided [major instructions for building jqpy on windows](https://github.com/mwilliamson/jq.py/issues/20#issuecomment-1784256183).

### CODE REVIEW

**Comment by mwilliamson** [comment]
Thanks for taking a look at this! I haven't had a chance to read through the changes yet, but I appreciate the thorough high-level description of the changes.

> MinGW GCC is used in above process. It has to be noted, that there does exist some possibilities of potential incompatibilities when the wheels are installed to python built with MSVC, but filling the blank of functional wheels for windows users should come with higher priority.

I'm torn on this point: while Windows wheels would be useful, providing them when they might not always work properly feels risky to me. Do we know what the potential incompatibilities are?

I'm not particularly familiar with how msys2 works: presumably it provides a libc implementation? Does that get included in the wheel? More generally, I'd be curious to know what ends up in the wheel.

Besides those questions, I think what might put my mind at ease would be to ask around in the Python packaging community to see if and how others have done similar things, and to get a sense of whether this approach is likely to be problematic or not.

**Comment by jknockel** [comment]
I'm very happy to see this. :) I have just one question: Am I correct in interpreting that you link against MSVCRT for 32-bit Windows and UCRT version 64-bit for Windows?  Why not link against MSVCRT for both?  I don't think we need anything from UCRT, and linking against UCRT might create hurdles for users of older versions of Windows which the versions of python we are currently targeting still support.

> I'm not particularly familiar with how msys2 works: presumably it provides a libc implementation? Does that get included in the wheel? More generally, I'd be curious to know what ends up in the wheel.

MinGW links against either the MSVCRT runtime (which has come with every version of Windows since Windows 95) or the UCRT runtime (which has come with every version of Windows since Windows 10 and can be optionally installed on Windows 7 SP1 and Windows 8.1). We statically link any other runtime dependency [here](https://github.com/mwilliamson/jq.py/blob/5c935ff5eba842fc99ff08a5f84711995910d525/setup.py#L91), so the end result is no *.dlls need to be distributed, although if we link against the UCRT runtime we do require that runtime to have already been installed.

> MinGW GCC is used in above process. It has to be noted, that there does exist some possibilities of potential incompatibilities when the wheels are installed to python built with MSVC, but filling the blank of functional wheels for windows users should come with higher priority.

Can you be more specific your concern?  For vanilla C (i.e., not C++) linking I think there is not any particular reason for concern, but perhaps I am not being imaginative enough here. :)

**Comment by du33169** [comment]
### MinGW vs MSVC

> MinGW GCC is used in above process. It has to be noted, that there does  exist some possibilities of potential incompatibilities when the wheels  are installed to python built with MSVC

This note actually refers to the sentence from the [WindowsCompilers page on python wiki](https://wiki.python.org/moin/WindowsCompilers#:~:text=Some%20alternative%20compilers%20exist):

> *Some alternative compilers exist like [MinGW](http://mingw.org/), but incompatibilities may occur with a CPython official distribution that is built with Microsoft Visual C++.* 

This statement is vague. Later I realized it might be saying that CPython itself may behave differently when compiled with other compilers.

> Do we know what the potential incompatibilities are?

In our situation, when we build a wheel, it generates a pyd file, which is also a dll file on windows. The problem is simplified to this: **"Are there potential incompatibilities loading a dll built with MinGW, from an MSVC-built executable(CPython)?"**

Most incompatibilities come from the differences in C++ ABI. Luckily, here we use pure C, and C ABI is consistent among compilers, so this is not a problem.

Apart from that, any program can properly load dll as long as the platform, calling conventions, and linking conventions match (according to [Microsoft docs](https://learn.microsoft.com/en-us/cpp/build/dlls-in-visual-cpp?view=msvc-170)). This kind of matching is ensured by MinGW. In other words, a dll is self-contained, and we don't need to care about what compiler is used to generate it. 

It is for sure that MinGW handles something differently from MSVC, but these differences are limited within the dll boundary unless we [share CRT objects across it](https://learn.microsoft.com/en-us/cpp/c-runtime-library/potential-errors-passing-crt-objects-across-dll-boundaries?view=msvc-170)(for example, allocate memory in dll and free it in the main program). I didn't examine all jq interfaces, but I suppose that is beyond what a python extension can do and CPython will not allow it.

> I'm not particularly familiar with how msys2 works: presumably it  provides a libc implementation? Does that get included in the wheel? 

This [MSYS2 wiki page](https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/) will help answer this question. MSYS2 is just an environment providing POSIX-compatible tools(make, autoconf, etc) for building **native** Windows software. Some of its tools might require a POSIX compatibility layer(`msys-2.0.dll`), but their built binaries do not.  

MinGW gcc finally links to the native C runtime (MSVCRT or UCRT) (dynamically), which is part of the Windows system. Once compiled, the binaries become native windows ones and have no longer anything to do with MSYS2 (unless dynamically linked to any dll used only by MSYS2).

> More generally, I'd be curious to know what ends up in the wheel.

Some extra libraries are needed to make lib jq work on Windows, as specified in `extra_link_args`: `-Wl,-Bstatic -lpthread -lshlwapi -static-libgcc`(from original @jknockel's fork):

- pthread: Windows has no native support for pthread. Mingw-w64 provides a compatible `libwinpthread-1.dll` (package `mingw-w64-ucrt-x86_64-winpthreads` in MSYS2). We have to statically link it.
- [shlwapi](https://learn.microsoft.com/en-us/windows/win32/api/shlwapi/): Windows provides this dll, so it actually doesn't need to be(and actually cannot be) static linked.
- [libgcc](https://gcc.gnu.org/onlinedocs/gccint/Libgcc.html): This seems to have no effect on Windows, and everything works fine without it. Is it some kind of fallback?

### build python extension with MinGW

> Besides those questions, I think what might put my mind at ease would be to ask around in the Python packaging community to see if and how  others have done similar things, and to get a sense of whether this  approach is likely to be problematic or not.

Building python extensions using mingw for MSVC python is an [old issue](https://github.com/python/cpython/issues/48959), especially for scientific extensions due to their dependency on Fortran, which MSVC doesn't support. It seems not a problem anymore, since mingw-w64 has developed well and become the major successor of the old mingw.

For example, scipy uses mingw gcc in msys2 (from [Rtools](https://cran.r-project.org/bin/windows/Rtools/rtools44/rtools.html)) to build their win_amd64 wheels (see [their workflow file](https://github.com/scipy/scipy/blob/de85b05990a475bfe04bd5f113511c4e9c2e2c8f/.github/workflows/wheels.yml#L111-L116)). Numpy supports [building it from source](https://numpy.org/devdocs/building/index.html#system-level-dependencies) using the same toolchain if the user requires f2py.

### MSVCRT or UCRT

> I have just one question: Am I correct in interpreting that you link  against MSVCRT for 32-bit Windows and UCRT version 64-bit for Windows?   Why not link against MSVCRT for both?
>   I don't think we need anything from UCRT, and linking against UCRT  might create hurdles for users of older versions of Windows which the  versions of python we are currently targeting still support.

Currently, [CPython uses VS2017](https://github.com/python/cpython/blob/78634cfa3dd4b542897835d5f097604dbeb0f3fd/PCbuild/readme.txt#L61-L62), linking to UCRT by default (both 32-bit and 64-bit). Their installer will [ship an app-local UCRT runtime](https://github.com/python/cpython/issues/76688) when installing python on a system without UCRT. So the distribution should not be a problem. UCRT is also currently recommended by Microsoft, and it should bring better support and compatibility for our win_amd64 wheels. 

In the current workflow, lib jq is built in MSYS2 using `mingw-w64-i686/ucrt-x86_64-gcc`, but when building the final wheel I only assign `CC=mingw-w64-i686-gcc` [if the target platform is win32](https://github.com/mwilliamson/jq.py/blob/5c935ff5eba842fc99ff08a5f84711995910d525/.github/workflows/build.yml#L212). For win_amd64, the final wheel is built using the standalone mingw gcc (version 12.2.0) pre-installed in the github running image (for simplicity), which also links to UCRT by default. We assume it should behave the same with `mingw-w64-ucrt-x86_64-gcc` in MSYS2(version 14.1.0). If this version mismatch is not acceptable, I would recommend we append `CC=mingw-w64-ucrt-x86_64-gcc` for consistency.

For 32-bit, [MSYS2 doesn't support ucrt32](https://github.com/msys2/MINGW-packages/issues/6901) and there is no `mingw-w64-ucrt-i686-gcc` package. So I can't directly link to UCRT despite I am willing to. However, there does exist some [standalone mingw i686 gcc executables support UCRT](https://github.com/niXman/mingw-builds-binaries/releases). So there are three choices:

1. Link to old MSVCRT for win32 wheels, which is the current behavior of this PR. This mix of CRT should be harmless as discussed earlier.
2. Modify the workflow and use standalone mingw i686 gcc that links to UCRT, but it increases complexity.
3. Don't provide official wheels for win32 at all, just like scipy (see [its pypi page](https://pypi.org/project/scipy/#files)).

**Comment by jknockel** [comment]
Thank you for the very detailed response!

> [libgcc](https://gcc.gnu.org/onlinedocs/gccint/Libgcc.html): This seems to have no effect on Windows, and everything works fine without it. Is it some kind of fallback?

It was necessary in my build environment to prevent a dynamic dependency on the *.dll.  If you are confident that it can be removed in yours (e.g., after removing it there is no dependency on a libgcc DLL in Dependency Walker), then I suppose statically linking against it can be removed.

> Link to old MSVCRT for win32 wheels, which is the current behavior of this PR. This mix of CRT should be harmless as discussed earlier.

Yes I think that this is fine.  In my builds, I linked against MSVCRT in the 64-bit version as well.

For purposes of jq.py, I don't think that there is any advantage to linking against MSVCRT versus the UCRT.  If anything, I was advocating for linking against MSVCRT in the 64-bit version as well for solely purposes of compatibility. We target python 3.6, which supports Windows as early as Windows Vista.  But, as you noted, python since at least 3.6 bundles the UCRT runtime, so we should have no problem with compatibility linking against UCRT either.

So, in other words, I don't think it really matters which we link against, so I would keep with the current solution if that is simplest!

**Comment by mwilliamson** [comment]
Thanks for your work and answers @du33169 and @jknockel. I've simplified the changes a little so that the Windows wheels are built as similarly as possible to the other wheels. I'll make a 1.8.0a1 release shortly to give an opportunity for the wheels to be tested.
